File: LintCode/cpp/Merge k Sorted Lists.cc
/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param lists: a list of ListNode
     * @return: The head of one sorted list.
     */
    ListNode *mergeKLists(vector<ListNode *> &lists) {
        priority_queue<ListNode*, 
                vector<ListNode*>, 
                function<bool(ListNode*, ListNode*)> > pq(
            [](const ListNode* a, const ListNode* b) -> bool {
                auto aa = a? a->val: INF;
                auto bb = b? b->val: INF;
                return aa > bb;
            });
        
        for (auto l: lists) {
            pq.push(l);
        }
        
        ListNode dummy_head(-1);
        ListNode* ptr = &dummy_head;
        
        while (!pq.empty()) {
            auto p = pq.top();
            pq.pop();
            
            if (p == nullptr) {
                continue;
            }
            ptr->next = p;
            ptr = ptr->next;
            
            p = p->next;
            pq.push(p);
        }
        ptr->next = nullptr;
        return dummy_head.next;
    }
private:
    static const int INF = 0x3f3f3f3f;
};

File: LintCode/python/Merge k Sorted Lists.py
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
import heapq

INF = 10 ** 10

class Solution:
    """
    @param lists: a list of ListNode
    @return: The head of one sorted list.
    """
    def mergeKLists(self, lists):
        pq = []
        dummyHead = ListNode(-1)
        tailNode = dummyHead
        for l in lists:
            heapq.heappush(pq, (-INF if l is None else l.val, l))
            
        while pq:
            (v, l) = heapq.heappop(pq)
            if l is None:
                continue
            tailNode.next = ListNode(v)
            tailNode = tailNode.next
            
            l = l.next
            heapq.heappush(pq, (-INF if l is None else l.val, l))
        
        return dummyHead.next

